### PR TITLE
Don't allocate a checker if the checker waiter is canceled.

### PR DESCRIPTION
**Pull request by MikaelMayer**
I'm wondering if this can help.
While debugging, I have seen a lot of checkers being allocated and prepared to just be thrown away because the task awaiting it has been canceled.
This PR does not even allocate the tasks.

However, it contains something I would like someone (e.g. @cpitclaudel ) to confirm it is not a problem: If we use WaitASync(cancellationToken), can it cause a deadlock in any way?

### CODE REVIEW

File: Source/VCGeneration/SplitAndVerifyWorker.cs [hunk 1]
```diff
@@ -98,10 +98,9 @@ private void TrackSplitsCost(List<Split> splits)
 
     async Task DoWork(Split split, CancellationToken cancellationToken)
     {
-      var checker = await split.parent.CheckerPool.FindCheckerFor(split.parent, split);
+      var checker = await split.parent.CheckerPool.FindCheckerFor(split.parent, split, cancellationToken).WaitAsync(cancellationToken);
```
Comment (keyboardDrummer) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:101]: Why did you add `.WaitAsync(cancellationToken)`? I think that's only useful when the Task doesn't have cancellation support, but that's what you're already adding in this PR

File: Source/VCGeneration/CheckerPool.cs [hunk 2]
```diff
@@ -11,7 +12,7 @@ namespace VC
   public class CheckerPool
   {
     private readonly ConcurrentStack<Checker> availableCheckers = new();
-    private readonly ConcurrentQueue<TaskCompletionSource<Checker>> checkerWaiters = new();
+    private readonly ConcurrentQueue<(TaskCompletionSource<Checker>, CancellationToken?)> checkerWaiters = new();
```
Comment (keyboardDrummer) [inline Source/VCGeneration/CheckerPool.cs:15]: I think you don't need to change the field, because you can cancel the `TaskCompletionSource` in the field already.

Here's an example of what I think you could do:

```csharp
  public class CheckerPool
  {
    private readonly ConcurrentStack<Checker> availableCheckers = new();
    private readonly ConcurrentQueue<TaskCompletionSource<Checker>> checkerWaiters = new();
    private int notCreatedCheckers;
    private bool disposed;

    public VCGenOptions Options { get; }

    public CheckerPool(VCGenOptions options)
    {
      this.Options = options;
      notCreatedCheckers = options.VcsCores;
    }

    public Task<Checker> FindCheckerFor(ConditionGeneration vcgen, Split split = null, CancellationToken? cancellationToken = null)
    {
      if (disposed) {
        return Task.FromException<Checker>(new Exception("CheckerPool was already disposed"));
      }
        
      if (availableCheckers.TryPop(out var result)) {
        PrepareChecker(vcgen.program, split, result);
        Contract.Assert(result != null);
        return Task.FromResult(result);
      }

      var afterDecrement = Interlocked.Decrement(ref notCreatedCheckers);
      if (afterDecrement >= 0) {
        var checker = CreateNewChecker();
        PrepareChecker(vcgen.program, split, checker);
        Contract.Assert(checker != null);
        return Task.FromResult(checker);
      }
      Interlocked.Increment(ref notCreatedCheckers);

      var source = new TaskCompletionSource<Checker>();
      cancellationToken?.Register(() =>
      {
        source.SetCanceled(cancellationToken.Value);
      });
      checkerWaiters.Enqueue(source);
      return source.Task.ContinueWith(task =>
      {
        PrepareChecker(vcgen.program, split, task.Result);
        Contract.Assert(task.Result != null);
        return task.Result;
      });
    }

    private Checker CreateNewChecker()
    {
      var log = Options.ProverLogFilePath;
      if (log != null && !log.Contains("@PROC@") && availableCheckers.Count > 0) {
        log = log + "." + availableCheckers.Count;
      } 

      return new Checker(this, log, Options.ProverLogFileAppend);
    }

    public void Dispose()
    {
      while (availableCheckers.TryPop(out var checker)) {
        checker.Close();
      }
      disposed = true;
    }

    void PrepareChecker(Program program, Split split, Checker checker)
    {
      if (checker.WillingToHandle(program) && (split == null || checker.SolverOptions.RandomSeed == split.RandomSeed && !Options.Prune))
      {
        checker.GetReady();
        return;
      }

      checker.Target(program, checker.TheoremProver.Context, split);
      checker.GetReady();
    }

    public void AddChecker(Checker checker)
    {
      if (checker.IsClosed) {
        throw new Exception();
      }
      if (disposed) {
        checker.Close();
        return;
      }

      while (checkerWaiters.TryDequeue(out var waiter)) {
        if (waiter.TrySetResult(checker)) {
          return;
        }
      }
      availableCheckers.Push(checker);
    }

    public void CheckerDied()
    {
      Checker newChecker = null;
      while (checkerWaiters.TryDequeue(out var waiter)) {
        if (waiter.Task.Status == TaskStatus.Running) {
          newChecker ??= CreateNewChecker();
          if (waiter.TrySetResult(newChecker)) {
            return;
          }
        }
      }

      if (newChecker != null) {
        AddChecker(newChecker);
      } else {
        Interlocked.Increment(ref notCreatedCheckers);
      }
    }
  }
```

**Comment by keyboardDrummer** [comment]
I got inspired by this PR and made one that also provides the cancellation feature but including a bugfix and refactoring: https://github.com/boogie-org/boogie/pull/552

**Comment by MikaelMayer** [comment]
Wonderful ! I was thinking that your PR is way better than this one as the refactoring is much cleaner. I'll close this PR.
